### PR TITLE
VRT 'if' option for vrt:// connection

### PR DESCRIPTION
**Pull request by mdsumner**
Adds 'if' (initial format) to the allowed options for a "vrt:// connection.

## Tasklist

 - [x] Add test case(s)
 - [x] Add documentation
 - [x] Review
 - [x] Adjust for comments
 - [ ] All CI builds and checks have passed

### CODE REVIEW

**Review by rouault — CHANGES_REQUESTED** [review]

File: frmts/vrt/vrtdataset.cpp [hunk 1]
```diff
@@ -1082,6 +1082,16 @@ GDALDataset *VRTDataset::OpenVRTProtocol(const char *pszSpec)
                     argv.AddString(aosGCP[j]);
                 }
             }
+            else if (EQUAL(pszKey, "if"))
+            {
+                CPLStringList aosInitialFormats(
```
[the remark this answers is not in the document]
Comment (mdsumner) [reply frmts/vrt/vrtdataset.cpp:1087]: awesome thank you

[the remark this answers is not in the document]
Comment (mdsumner) [reply frmts/vrt/vrtdataset.cpp]: I've done this now, left a note in the OP

File: autotest/gcore/vrt_read.py [hunk 2]
```diff
@@ -1478,6 +1478,15 @@ def test_vrt_protocol():
     ds = gdal.Open("vrt://data/float32.tif?gcp=invalid")
     assert ds is None
 
+    ds = gdal.Open("vrt://data/float32.tif?if=AAIGrid,COG")
```
[the remark this answers is not in the document]
Comment (mdsumner) [reply autotest/gcore/vrt_read.py]: I've got better tests now :)

File: frmts/vrt/vrtdataset.cpp [hunk 3]
```diff
@@ -954,15 +954,61 @@ GDALDataset *VRTDataset::OpenVRTProtocol(const char *pszSpec)
...
         osFilename.resize(nPosQuotationMark);
     }
-    auto poSrcDS = GDALDataset::Open(
-        osFilename, GDAL_OF_RASTER | GDAL_OF_SHARED, nullptr, nullptr, nullptr);
+
+    // Parse query string, get args required for initial Open()
+    CPLStringList aosTokens0(CSLTokenizeString2(osQueryString, "&", 0));
+    char **papszAllowedDrivers = nullptr;
+    int iRemoveOpt;  // not allowed to have multiple &if=drv1&if=drv2
+    iRemoveOpt = -1;
```
Comment (rouault) [inline frmts/vrt/vrtdataset.cpp:962]:
```suggestion
    CPLStringList aosAllowedDrivers;
    int iRemoveOpt = -1;  // not allowed to have multiple &if=drv1&if=drv2
```

File: frmts/vrt/vrtdataset.cpp [hunk 4]
```diff
@@ -954,15 +954,61 @@ GDALDataset *VRTDataset::OpenVRTProtocol(const char *pszSpec)
...
+                }
+                CPLStringList aosInitialFormats(
+                    CSLTokenizeString2(pszValue, ",", 0));
+                for (int j = 0; j < aosInitialFormats.size(); j++)
+                {
+                    CPLDebug("VRT", "Construct papszAllowedDrivers, add: %s\n",
+                             aosInitialFormats[j]);
+                    papszAllowedDrivers =
+                        CSLAddString(papszAllowedDrivers, aosInitialFormats[j]);
+                }
```
Comment (rouault) [inline frmts/vrt/vrtdataset.cpp:985]:
```suggestion
                aosAllowedDrivers = CSLTokenizeString2(pszValue, ",", 0);
```

File: frmts/vrt/vrtdataset.cpp [hunk 5]
```diff
@@ -954,15 +954,61 @@ GDALDataset *VRTDataset::OpenVRTProtocol(const char *pszSpec)
...
+        }
+        else
+        {
+            CPLDebug("VRT", "Removed 'if' arg at position: %i\n", iRemoveOpt);
+        }
+    }
+
+    auto poSrcDS =
+        GDALDataset::Open(osFilename, GDAL_OF_RASTER | GDAL_OF_SHARED,
+                          papszAllowedDrivers, nullptr, nullptr);
```
Comment (rouault) [inline frmts/vrt/vrtdataset.cpp:1006]:
```suggestion
                          aosAllowedDrivers.List(), nullptr, nullptr);
```

File: frmts/vrt/vrtdataset.cpp [hunk 6]
```diff
@@ -954,15 +954,61 @@ GDALDataset *VRTDataset::OpenVRTProtocol(const char *pszSpec)
...
+        const char *pszValue = CPLParseNameValue(aosTokens0[i], &pszKey);
+        if (pszKey)
+        {
+            if (EQUAL(pszKey, "if"))
+            {
+                if (iRemoveOpt > -1)
+                {
+                    CPLError(CE_Failure, CPLE_IllegalArg,
+                             "Invalid vrt:// 'if' format, must not occur "
+                             "multiple times (use comma separated list)\n");
```
Comment (rouault) [inline frmts/vrt/vrtdataset.cpp:975]:
```suggestion
                    CPLError(CE_Failure, CPLE_IllegalArg,
                             "Invalid vrt:// 'if' format, must not occur "
                             "multiple times (use comma separated list)\n");
                    return nullptr;
```

**Comment by rouault** [comment]
you also need to solve the conflict in vrt.rst

**Review by rouault — APPROVED** [review]

**Comment by mdsumner** [comment]
thanks so much, glad to get this one in - appreciate all the help :) learnt a lot from this one

**Pull request merged**